### PR TITLE
Ackermann Steering Plugin

### PR DESCRIPTION
**Pull request by knoedler**
## Summary
This adds an Ackermann Steering Plugin based on the differential drive plugin.

The plugin accepts a linear velocity (x) and a rotational velocity (z) the same as the differential drive plugin and sets the vehicle drive and steering joints to achieve the requested inputs.  The plugin and integration tests are largely based on the differential drive plugin.

Items to consider:
SpeedLimiter.cc and SpeedLimiter.hh are identical copies from the diff_drive plugin
Odometry is not as accurate as a pure diff_drive robot.  I had to open the integration tests limit up from 1cm to 25cm when comparing odometry to ground truth in the integration tests.  I tried various ground friction coefficients and using the wheel
slip plugin, but I did not achieve 1cm accuracy for the odometry.
There is a lot of commonality between the differential and Ackermann plugins.  It might make sense to merge them at some point in the future when the Ackermann plugin has aged a bit and is stable (to minimize risk to the heavily used diff drive plugin).
I was not able to generate a coverage report.  

## Test it
The plugin can be tested using examples/worlds/ackermann_steering.sdf

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example world and/or tutorial
- [x] Updated documentation (as needed)
- [NA] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [?] While waiting for a review on your PR, please help review

### CODE REVIEW

File: src/systems/ackermann_steering/AckermannSteering.cc [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+
+
+  // Subscribe to commands
+  std::vector<std::string> topics;
+  if (_sdf->HasElement("topic"))
+  {
+    topics.push_back(_sdf->Get<std::string>("topic"));
+  }
+  topics.push_back("/model/" + this->dataPtr->model.Name(_ecm) + "/cmd_vel");
+  auto topic = validTopic(topics);
```
Comment (ahcorde) [inline src/systems/ackermann_steering/AckermannSteering.cc:327]: You should check if the output of validTopic is empty. https://ignitionrobotics.org/api/transport/9.0/classignition_1_1transport_1_1TopicUtils.html#a42548e276fce157d9794eb8a05bf07b3

here and below

File: src/systems/ackermann_steering/AckermannSteering.cc [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+
+  double odomFreq = _sdf->Get<double>("odom_publish_frequency", 50).first;
+  if (odomFreq > 0)
+  {
+    std::chrono::duration<double> odomPer{1 / odomFreq};
+    this->dataPtr->odomPubPeriod =
+      std::chrono::duration_cast<std::chrono::steady_clock::duration>(odomPer);
+  }
+
+
```
Comment (ahcorde) [inline src/systems/ackermann_steering/AckermannSteering.cc:319]:
```suggestion
```

File: src/systems/ackermann_steering/AckermannSteering.cc [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+    }
+  }
+  else
+  {
+    auto childFrame = msg.mutable_header()->add_data();
+    childFrame->set_key("child_frame_id");
+    childFrame->add_value(this->sdfChildFrameId);
+  }
+
+
```
Comment (ahcorde) [inline src/systems/ackermann_steering/AckermannSteering.cc:673]:
```suggestion
```

File: src/systems/ackermann_steering/AckermannSteering.cc [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+  double phi = atan(this->wheelBase / turningRadius);
+
+  // Partially simulate a simple differential
+  this->rightJointSpeed =
+      (linVel * (1.0 + (this->wheelSeparation * tan(phi)) /
+                 (2.0 * this->wheelBase))) / this->wheelRadius;
+  this->leftJointSpeed =
+      (linVel * (1.0 - (this->wheelSeparation * tan(phi)) /
+                 (2.0 * this->wheelBase))) / this->wheelRadius;
+
```
Comment (ahcorde) [inline src/systems/ackermann_steering/AckermannSteering.cc:731]:
```suggestion
```

File: src/systems/ackermann_steering/AckermannSteering.cc [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+      this->rightSteeringJoints[0]);
+
+  // Abort if the joints were not found or just created.
+  if (!leftSteeringPos || !rightSteeringPos ||
+      leftSteeringPos->Data().empty() ||
+      rightSteeringPos->Data().empty())
+  {
+    return;
+  }
+
```
Comment (ahcorde) [inline src/systems/ackermann_steering/AckermannSteering.cc:745]:
```suggestion
```

File: src/systems/ackermann_steering/SpeedLimiter.cc [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+/// \brief Private SpeedLimiter data class.
+class ignition::gazebo::systems::SpeedLimiterPrivate
+{
+  /// \brief Class constructor.
+  public: SpeedLimiterPrivate(bool _hasVelocityLimits,
```
Comment (ahcorde) [inline src/systems/ackermann_steering/SpeedLimiter.cc:52]: please document each argument

File: test/integration/ackermann_steering_system.cc [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+  EXPECT_LT(poses[0].Rot().Z(), poses[3999].Rot().Z());
+}
+
+/////////////////////////////////////////////////
+TEST_P(AckermannSteeringTest, OdomFrameId)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/ackermann_steering.sdf");
```
Comment (ahcorde) [inline test/integration/ackermann_steering_system.cc:324]: please use `common::joinPaths`.

File: test/integration/ackermann_steering_system.cc [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+      "/model/foo/cmdvel", "/model/bar/odom");
+}
+
+/////////////////////////////////////////////////
+TEST_P(AckermannSteeringTest, SkidPublishCmd)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/ackermann_steering_slow_odom.sdf");
```
Comment (ahcorde) [inline test/integration/ackermann_steering_system.cc:226]: please use `common::joinPaths` to allow interoperability between OS

File: test/integration/ackermann_steering_system.cc [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+  EXPECT_EQ(5u, odomPosesCount);
+}
+
+/////////////////////////////////////////////////
+TEST_P(AckermannSteeringTest, OdomCustomFrameId)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/ackermann_steering_custom_frame_id.sdf");
```
Comment (ahcorde) [inline test/integration/ackermann_steering_system.cc:382]: please use `common::joinPaths`

File: src/systems/ackermann_steering/AckermannSteering.hh [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  /// of left_joint, right_joint, left_steering_joint and
+  /// right_steering_joint
+  ///
+  /// References:
+  /// https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo4/src/systems/diff_drive
+  /// https://www.auto.tuwien.ac.at/bib/pdf_TR/TR0183.pdf
+  /// https://github.com/froohoo/ackermansteer/blob/master/ackermansteer/
+
+
+
```
Comment (ahcorde) [inline src/systems/ackermann_steering/AckermannSteering.hh:96]:
```suggestion

```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/systems/ackermann_steering/AckermannSteering.cc (hunk 1 above)
Comment (knoedler) [reply src/systems/ackermann_steering/AckermannSteering.cc:326]: Thank you for looking at the code.  All requested changes have been made and checked in.  codecheck and tests have been re-run and pass.
https://github.com/ignitionrobotics/ign-gazebo/pull/618/commits/7d56600edb0dd7d9695ee207a4f98a177f8b414e#diff-0dcac5ee6e5c74605877efaf6a30d6c6b5b23ef9d867f238e7a4b5185f1c4de0R327-R331

File: test/integration/ackermann_steering_system.cc (hunk 7 above)
Comment (knoedler) [reply test/integration/ackermann_steering_system.cc]: I addressed the three you found plus two others.
https://github.com/ignitionrobotics/ign-gazebo/pull/618/commits/7d56600edb0dd7d9695ee207a4f98a177f8b414e#diff-0dcac5ee6e5c74605877efaf6a30d6c6b5b23ef9d867f238e7a4b5185f1c4de0R327-R331

File: src/systems/ackermann_steering/SpeedLimiter.cc (hunk 6 above)
Comment (knoedler) [reply src/systems/ackermann_steering/SpeedLimiter.cc:61]: https://github.com/ignitionrobotics/ign-gazebo/pull/618/commits/7d56600edb0dd7d9695ee207a4f98a177f8b414e#diff-30d8d7c4b4f1b1762680c8a1895cb99472a7ddd049953fd6fab043d9d09157acR52-R60

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Review by chapulina — COMMENTED** [review]
This is looking good to me! I didn't check the math too closely though. I took a quick look at the implementation on [gazebo_ros_pkgs](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/50c2d1b329936b2c60cdd78aabbf616ff691e43c/gazebo_plugins/src/gazebo_ros_ackermann_drive.cpp#L513-L534) but the approach looks a bit different.

**Review by ahcorde — CHANGES_REQUESTED** [review]
Remember to sign all your commits. DCO is failing

**Comment by knoedler** [comment]
> Remember to sign all your commits. DCO is failing

That particular commit was from clicking on "commit suggestion".  I would have expected github to handle signoffs.  I have made a couple of attempts to resolve it, but my googling has not found a successful way to resolve it.

I believe I have addressed the issues from the two reviews.

**Comment by chapulina** [comment]
>  I would have expected github to handle signoffs. 

Yeah unfortunately GitHub doesn't make it easy to sign commits from the web UI. We've been manually adding the signature for those cases.

One way of solving this would be by adding the signature locally while rebasing and then force-pushing. If this is too much trouble, we can manually accept the DCO check since all other commits are properly signed.

**Comment by knoedler** [comment]
> 
> One way of solving this would be by adding the signature locally while rebasing and then force-pushing. If this is too much trouble, we can manually accept the DCO check since all other commits are properly signed.

If you could, please manually accept the DCO check.

Adding the signature locally/rebasing/force-pushing was one of the approaches I tried a couple of different ways, but it didn't work for me in this case.

**Review by chapulina — APPROVED** [review]
Thank you for the new plugin!

**Pull request merged**